platform/src/main/python/dlpx/virtualization/platform/_upgrade.py

```diff
@@ -95,28 +113,26 @@ def _success_upgrade_response(upgraded_dict):
             return_value=upgrade_result)
         return upgrade_response
 
-    def __process_upgrade_request(self, request, id_to_impl):
-        """Iterate through all objects in the pre_upgrade_parameters map,
-        invoke all available migrations on each object and its metadata,
-        and return a map containing the updated metadata for each object.
+    @staticmethod
+    def _run_migration_upgrades(request, lua_impls_getter,
+                                platform_impls_getter):
+        """
+        Given the list of lua and platform migration to run, iterate and
+        invoke these migrations on each object and its metadata, and return a
+        dict containing the upgraded parameters.
         """
         post_upgrade_parameters = {}
+        impls_list = lua_impls_getter(
```

Comment:
Suppose we have this situation:
Lua toolkits 1.0, 2.0, and 3.0 released.

Python plugin 4.0 released. For this plugin the plugin author did not feel like bothering to write Python migrations for 1.0. So, this only includes lua migration IDs `2.0` and `3.0`.

Suppose the 4.0 plugin is uploaded to an engine with the 1.0 toolkit (only) installed. In that case, the engine will send over a protobuf message with `lua_upgrade_version` set to "1.0".

What happens in this case?  It seems like we should error out, but I think this code will just run the "2.0" and "3.0" migrations.
- Maybe I'm missing something where we check for this case?
- Maybe we're relying on the schema check at the very end? If that doesn't fail, then apparently everything worked and there was no need for a "1.0" migration?
- Maybe we're expecting that the engine should be checking for this, and not ever even sending a Protobuf message if it detects this case?

Reply:
yeah i think the idea was that if there was something missing, the schema checks would fail. I wasn't sure if I wanted to add a specific error because even though we technically expect when a major version increases that there was a schema change I don't believe (and I could be wrong) that we actually require that there is a schema change.

Oh the other hand this is relevant to my reply to one of jeff's comments as well as something I had thought about. Depending on if we create a list of lua migrations that are implemented, we may be able to add a check on the delphix engine to not send over the upgrade request at all. but again I wasn't sure if we wanted to restrict that if for some reason no schema was actually needed (obviously work around would be to write migration that does nothing. either way I think we just need to decide and document)

Reply:
One thing we should consider with this is user error messages.  If the schema check fails, then they'll get an error that suggests a bug in the plugin. We might be able to reword that message to say "or maybe this Lua version isn't supported"

But also, any of the migrations might fail.

Imagine in my example that 1.0->2.0 adds a new property called `port`. Then 2.0->3.0 changes the name of that property to `databasePort`:
```
@plugin.upgrade.virtual_source("2.0", type=MigrationType.LUA)
def two_to_three(old_virtual_source):
  old_virtual_source["databasePort"] = old_virtual_source["port"
  del old_virtual_source["port"]
  return
```

So, now the upgrade will fail, but the error message will just report on a Python error like this:
```
Traceback (most recent call last):
  File "plugin_runner.py", line 73, in <module>
    old_virtual_source["databasePort"] = old_virtual_source["port"]
KeyError: 'port'
```

This would also suggest a bug in the plugin, when really it's just that the Lua version isn't supported.

Reply:
The team talked offline about this and have decided that we will add a new field in the plugin config = lowestSupportedLuaVersion which the plugin author will have to provide. We filed a new bug for the sdk to add this (https://github.com/delphix/virtualization-sdk/issues/115) as well as the delphix engine to pull it in and do the actually check (https://jira.delphix.com/browse/PYT-1338).